runtime/compiler/trj9/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -20927,6 +20932,81 @@ J9::Z::TreeEvaluator::pdstoreVectorEvaluatorHelper(TR::Node *node, TR::CodeGener
    return NULL;
    }
 
+/**
+ * \brief Used by pdstore and pdload vector evaluators to fold constant address displacements into
+ * vector load and store instructions.
+ *
+ * \details
+ *
+ * Vector loads and stores use displacement and base register to reference memory locations. When the
+ * memory reference has no index registers, the displacement constant can be folded into the vector load/store
+ * instruction. This applies to vector load and store instructions such as VSTRL, VSTEB, VSTEH, VLRL, and etc.
+ *
+ * For example, the following instruction sequence has no constant folding:
+ *
+ * \verbatim
+ *
+ * LA       GPR12,X'8'(,GPR1)
+ * VSTRL   VR0,X'0'(GPR12),0
+ *
+ * \endverbatim
+ *
+ * The LA instruction can be eliminated if the constant is folded into the store instruction, resulting in:
+ *
+ * \verbatim
+ * VSTRL   VR0,X'8'(GPR1),0
+ * \endverbatim
+ *
+ */
+TR::Register *
+J9::Z::TreeEvaluator::foldAddressDisplacementIntoVectorLoadStore(TR::Node* addressNode,
```

Comment:
Why is this function needed at all? There already exists logic in `generateMemoryReference` API to handle folding of the displacement constant into the memref. In the very first deleted lines in this PR:

```
TR::Register* addressReg = cg->evaluate(addressNode);
TR::MemoryReference* sourceMR = generateS390MemoryReference(addressReg, NULL, 0, cg);
```

Instead of doing this simply constructing the memory reference by passing in the `addressNode` will result in no `LA` instruction being generated. See the following API:

https://github.com/eclipse/omr/blob/225a9d93e5e202446df9dc368b04e567c48cbe34/compiler/z/codegen/OMRMemoryReference.hpp#L467

This is how memory references are created for all generic loads without materializing the memory reference. If code does not exist to handle this particular case then logic needs to be added in the MemoryReference constructor where all the other folding logic is.

Reply:
Changed the way we create memory reference to use the API you pointed out.
This change depends on an OMR change: https://github.com/eclipse/omr/pull/2391